File: static/js/collector.js
class BehavioralDataCollector {
    constructor(apiEndpoint) {
        this.apiEndpoint = apiEndpoint;
        this.sessionId = this._generateSessionId();
        this.keystrokeBuffer = [];
        this.mouseBuffer = [];
        this.sessionStart = Date.now();
        this.bufferSize = 50; // Send data when buffer reaches this size
        this.setupEventListeners();
        console.log('BehavioralDataCollector initialized with session ID:', this.sessionId);
    }

    _generateSessionId() {
        return Math.random().toString(36).substring(2, 15) +
               Math.random().toString(36).substring(2, 15);
    }

    setupEventListeners() {
        // Keyboard events
        document.addEventListener('keydown', (e) => this._handleKeystroke(e, 'keydown'));
        document.addEventListener('keyup', (e) => this._handleKeystroke(e, 'keyup'));

        // Mouse events
        document.addEventListener('mousemove', (e) => this._handleMouseMove(e));
        document.addEventListener('click', (e) => this._handleMouseClick(e));
        
        // Send data periodically and before page unload
        setInterval(() => this.sendData(), 5000);
        window.addEventListener('beforeunload', () => this.sendData());
    }

    _handleKeystroke(event, eventType) {
        // Don't collect actual keys for privacy, just timing and key codes
        const keystrokeData = {
            event_type: eventType,
            key_code: event.keyCode,
            timestamp: Date.now() / 1000
        };

        this.keystrokeBuffer.push(keystrokeData);
        if (this.keystrokeBuffer.length >= this.bufferSize) {
            this.sendData();
        }
    }

    _handleMouseMove(event) {
        const mouseData = {
            type: 'movement',
            x: event.clientX,
            y: event.clientY,
            timestamp: Date.now() / 1000
        };

        this.mouseBuffer.push(mouseData);
        if (this.mouseBuffer.length >= this.bufferSize) {
            this.sendData();
        }
    }

    _handleMouseClick(event) {
        const clickData = {
            type: 'click',
            x: event.clientX,
            y: event.clientY,
            button: event.button,
            timestamp: Date.now() / 1000
        };

        this.mouseBuffer.push(clickData);
        this.sendData();
    }

    _collectDeviceInfo() {
        return {
            screen_resolution: `${window.screen.width}x${window.screen.height}`,
            browser_language: navigator.language,
            timezone_offset: new Date().getTimezoneOffset(),
            platform: navigator.platform,
            is_mobile: /Mobile|Android|iPhone/i.test(navigator.userAgent)
        };
    }

    async sendData() {
        if (this.keystrokeBuffer.length === 0 && this.mouseBuffer.length === 0) {
            return;
        }

        console.log('Sending behavioral data...');
        const sessionData = {
            session_id: this.sessionId,
            timestamp: new Date().toISOString(),
            session_duration: (Date.now() - this.sessionStart) / 1000,
            keystroke_data: [...this.keystrokeBuffer],
            mouse_data: [...this.mouseBuffer],
            device_info: this._collectDeviceInfo()
        };

        try {
            console.log('Sending data to server:', sessionData);
            const response = await fetch(`${this.apiEndpoint}/collect`, {
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json',
                },
                body: JSON.stringify(sessionData)
            });

            if (response.ok) {
                console.log('Data sent successfully');
                // Clear buffers only if data was successfully sent
                this.keystrokeBuffer = [];
                this.mouseBuffer = [];
            } else {
                console.error('Failed to send behavioral data:', await response.text());
            }
        } catch (error) {
            console.error('Error sending behavioral data:', error);
        }
    }

    async analyzeSession() {
        console.log('Analyzing session:', this.sessionId);
        try {
            const response = await fetch(`${this.apiEndpoint}/analyze`, {
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json',
                },
                body: JSON.stringify({
                    session_id: this.sessionId
                })
            });

            if (response.ok) {
                const result = await response.json();
                console.log('Analysis result:', result);
                return result;
            } else {
                const errorText = await response.text();
                console.error('Failed to analyze session:', errorText);
                return null;
            }
        } catch (error) {
            console.error('Error analyzing session:', error);
            return null;
        }
    }
}

// Example usage:
// const collector = new BehavioralDataCollector('http://localhost:5000');
// Later, you can analyze the session:
// const analysis = await collector.analyzeSession(); 

File: src/data_collection/collector.py
import time
from datetime import datetime
from typing import Dict, List, Any
import json
import hashlib

class BehavioralDataCollector:
    def __init__(self):
        self.keystroke_buffer: List[Dict[str, Any]] = []
        self.mouse_buffer: List[Dict[str, Any]] = []
        self.session_start = datetime.now()
        self.session_id = self._generate_session_id()

    def _generate_session_id(self) -> str:
        """Generate a unique session ID based on timestamp and random data"""
        timestamp = str(time.time()).encode()
        return hashlib.sha256(timestamp).hexdigest()[:16]

    def collect_keystroke(self, event_type: str, key_code: int, timestamp: float) -> None:
        """
        Collect keystroke events
        :param event_type: 'keydown' or 'keyup'
        :param key_code: ASCII code of the key
        :param timestamp: Unix timestamp of the event
        """
        self.keystroke_buffer.append({
            "event_type": event_type,
            "key_code": key_code,  # Store only the key code, not the actual key for privacy
            "timestamp": timestamp
        })

    def collect_mouse_movement(self, x: float, y: float, timestamp: float) -> None:
        """
        Collect mouse movement data
        :param x: X coordinate
        :param y: Y coordinate
        :param timestamp: Unix timestamp of the event
        """
        self.mouse_buffer.append({
            "type": "movement",
            "x": x,
            "y": y,
            "timestamp": timestamp
        })

    def collect_mouse_click(self, x: float, y: float, button: str, timestamp: float) -> None:
        """
        Collect mouse click events
        :param x: X coordinate
        :param y: Y coordinate
        :param button: Which mouse button was clicked
        :param timestamp: Unix timestamp of the event
        """
        self.mouse_buffer.append({
            "type": "click",
            "x": x,
            "y": y,
            "button": button,
            "timestamp": timestamp
        })

    def collect_device_info(self) -> Dict[str, Any]:
        """
        Collect anonymous device information
        Returns a dictionary of device characteristics
        """
        # Note: In a real implementation, this would use JavaScript to collect
        # browser/device information securely
        return {
            "screen_resolution": "1920x1080",  # Example
            "browser_language": "en-US",
            "timezone_offset": -240,  # Example: EST
            "platform": "MacOS",
            "is_mobile": False
        }

    def get_session_data(self) -> Dict[str, Any]:
        """
        Get all collected data for the current session
        """
        return {
            "session_id": self.session_id,
            "timestamp": datetime.now().isoformat(),
            "session_duration": (datetime.now() - self.session_start).total_seconds(),
            "keystroke_data": self.keystroke_buffer,
            "mouse_data": self.mouse_buffer,
            "device_info": self.collect_device_info()
        }

    def clear_buffers(self) -> None:
        """Clear all data buffers"""
        self.keystroke_buffer.clear()
        self.mouse_buffer.clear()

    def save_session(self, filepath: str) -> None:
        """
        Save the current session data to a file
        :param filepath: Path to save the session data
        """
        with open(filepath, 'w') as f:
            json.dump(self.get_session_data(), f, indent=2) 